### DISCUSSION TITLE
SML pulse counter on "slow" pulses

### DISCUSSION THREAD

**Discussion by unciokas** [arendst/Tasmota#18171] [Q&A] [answered]
I'm counting pulses from electricity meter. In tasmota 12.3.1 everything was ok, after upgrade to 12.4 total count is OK, and usage shows good only when pulse rate is less than 5s. If pulse rate is lower than 5s it shows some random value (1327W) on received pulse, than shows 0W after 5s and than again random value on next pulse. Total count is OK.

>D

>M 1
+1,8,c,1,-20,Power
1,1-0:1.7.0*255(@0.01667,power,W,Usage,0)
1,1-0:1.8.0*255(@1000,consumption,KWh,Total,3)
#

**Comment by gemu2015** [comment, marked as answer]
this timeout was on request from another user.
#define CNT_PULSE_TIMEOUT 5000

i will make this overridable in the next pr.

so in the meantime search for this define in the source code and change as requested.

**Reply by dr-knor** [reply]
Hello gemu2015, has the timeout already been made overridable in one of the subsequent releases? I have the same issue that unciokas describes. A timeout of 5000 ms is quite short. On my (rather standard I believe) kWh counter of 1 pulse per Wh this amounts to the counter dropping out below 720W. Since this counter monitors a small PV setup (510 Wp) this renders the power measurement function useless for my application. A timeout of (up to) 1 hour (1W in my case) would be more practical.

**Reply by gemu2015** [reply]
#define CNT_PULSE_TIMEOUT xxx  defaults to 5000

**Reply by esik007** [reply]
Hi all, 
#define CNT_PULSE_TIMEOUT setting not solve problem. Increase value decrease faulting power level, but under this level program raise wrong random values. It is in Tasmota-13.3.0/tasmota/tasmota_xsns_sensor/xsns_53_sml.ino. 
The solution would be to introduce another parameter that would reset the output when the limit CNT_PULSE_TIMEOUT is reached. So if the pulse length is greater than or equal to the set value, the output 1-0:1.7.0255 would return 0W.

**Reply by gemu2015** [reply]
but that is exactly what this timer does. it resets the output after this timeout				

```
if ((time - sml_counters[cindex].sml_counter_lfalltime) > CNT_PULSE_TIMEOUT) {
 InjektCounterValue(meters, RtcSettings.pulse_counter[cindex], 0);
 sml_counters[cindex].sml_counter_lfalltime = time;
}
```

**Reply by esik007** [reply]
OK. I don't read source.
In this case, it seems to me, do this: in CNT_PULSE_TIMEOUT isn't pulse. Program correctly reset output. Run next timeout and delayed pulse will arrive. Program calculate power, but wrong value. Solution - in case reset, start measuring power just by receiving next pulse.

**Reply by dr-knor** [reply]
@gemu2015 Thanks! I added #define CNT_PULSE_TIMEOUT 3600000 to user_config_override.h before compile, and the power (W) readings now appear to be good. Also the Wh counting was working OK during the day (as observed via Domoticz). The only strange thing that happened is that at the end of the day the total energy generation count (from 1,1-0:1.8.0*255(@1,generation,Wh,Total_in,0) is back at zero in the Tasmota main menu.  I am not sure when exactly that happened and if it happened only once or multiple times. I am not sure if it could be related.

**Reply by dr-knor** [reply]
Few hours later... the Total_in count number is back again on the main menu, with what seems to be the correct value (starting value of today + energy generated today). So apparently the Total_in number was not lost, but Tasmota was just temporarily printing zero on the main menu for some reason.

**Comment by unciokas** [comment]
OK, that solves my problem.
Thank You for fast answer.

**Comment by esik007** [comment]
Hi qemu2015. I have same problem. CNT_PULSE_TIMEOUT I increase to 20000. But it solve problem to faulty power level to 90W in my case ( 2000 imp/kWh ). Lower power raise random values. Is possible set the output power to 0W when the time between pulses is greater than 20s? Thanks

**Comment by esik007** [comment]
Hi, the problem with the power calculation error over the time between pulses more than CNT_PULSE_TIMEOUT (e.g. 20000W on PV 2800W) was finally solved by ChatGPT. I removed the line in the program code that caused this error, and the SO meter works correctly... 
What happens next (error scenario):
No pulses for a long time (e.g., PV system output ≈ 90 W)
A timeout occurs → power = 0 (correct)
sml_counter_lfalltime is set to “now”
The next pulse arrives
In this code:
sml_counters[cindex].sml_counter_pulsewidth =
    ctime - sml_counters[cindex].sml_counter_lfalltime;
the result is:
pulsewidth = a very small value (e.g., a few milliseconds)
calculation:
60000 / 3 ms ≈ 20,000 W

**Reply by esik007** [reply]
Commented line: 
if ((time - sml_counters[cindex].sml_counter_lfalltime) > CNT_PULSE_TIMEOUT) {
 InjektCounterValue(meters, RtcSettings.pulse_counter[cindex], 0);
 //---> sml_counters[cindex].sml_counter_lfalltime = time;
}
